ASTNode.h

```diff
@@ -1,735 +1,794 @@
+public:
+    ASTChainStore(list_t nodes, PycRef<ASTNode> src)
+        : ASTNodeList(nodes, NODE_CHAINSTORE), m_src(std::move(src)) { }
+    
+    PycRef<ASTNode> src() const { return m_src; }
+
+private:
+    PycRef<ASTNode> m_src;
+};
+
+
+class ASTObject : public ASTNode {
+public:
+    ASTObject(PycRef<PycObject> obj)
+        : ASTNode(NODE_OBJECT), m_obj(std::move(obj)) { }
+
+    PycRef<PycObject> object() const { return m_obj; }
+
+private:
+    PycRef<PycObject> m_obj;
+};
+
+
+class ASTUnary : public ASTNode {
+public:
+    enum UnOp {
+        UN_POSITIVE, UN_NEGATIVE, UN_INVERT, UN_NOT
+    };
+
+    ASTUnary(PycRef<ASTNode> operand, int op)
+        : ASTNode(NODE_UNARY), m_op(op), m_operand(std::move(operand)) { }
+
+    PycRef<ASTNode> operand() const { return m_operand; }
+    int op() const { return m_op; }
+    virtual const char* op_str() const;
+
+protected:
+    int m_op;
+
+private:
+    PycRef<ASTNode> m_operand;
+};
+
+
+class ASTBinary : public ASTNode {
+public:
+    enum BinOp {
+        BIN_ATTR, BIN_POWER, BIN_MULTIPLY, BIN_DIVIDE, BIN_FLOOR, BIN_MODULO,
+        BIN_ADD, BIN_SUBTRACT, BIN_LSHIFT, BIN_RSHIFT, BIN_AND, BIN_XOR,
+        BIN_OR, BIN_LOG_AND, BIN_LOG_OR, BIN_MAT_MULTIPLY,
+        /* Inplace operations */
+        BIN_IP_ADD, BIN_IP_SUBTRACT, BIN_IP_MULTIPLY, BIN_IP_DIVIDE,
+        BIN_IP_MODULO, BIN_IP_POWER, BIN_IP_LSHIFT, BIN_IP_RSHIFT, BIN_IP_AND,
+        BIN_IP_XOR, BIN_IP_OR, BIN_IP_FLOOR, BIN_IP_MAT_MULTIPLY,
+    };
+
+    static BinOp getBinOpFromOperand(int operand) {
+        switch (operand) {
+        case 0:
+            return BIN_ADD;
+        case 1:
+            return BIN_AND;
+        case 2:
+            return BIN_FLOOR;
+        case 3:
+            return BIN_LSHIFT;
+        case 4:
+            return BIN_MAT_MULTIPLY;
+        case 5:
+            return BIN_MULTIPLY;
+        case 6:
+            return BIN_MODULO;
+        case 7:
+            return BIN_OR;
+        case 8:
+            return BIN_POWER;
+        case 9:
+            return BIN_RSHIFT;
+        case 10:
+            return BIN_SUBTRACT;
+        case 11:
+            return BIN_DIVIDE;
+        case 12:
+            return BIN_XOR;
+        case 13:
+            return BIN_IP_ADD;
+        case 14:
+            return BIN_IP_AND;
+        case 15:
+            return BIN_IP_FLOOR;
+        case 16:
+            return BIN_IP_LSHIFT;
+        case 17:
+            return BIN_MAT_MULTIPLY;
+        case 18:
+            return BIN_IP_MULTIPLY;
+        case 19:
+            return BIN_IP_MODULO;
+        case 20:
+            return BIN_IP_OR;
+        case 21:
+            return BIN_IP_POWER;
+        case 22:
+            return BIN_IP_RSHIFT;
+        case 23:
+            return BIN_IP_SUBTRACT;
+        case 24:
+            return BIN_IP_DIVIDE;
+        case 25:
+            return BIN_IP_XOR;
+        default:
+            return BIN_ADD; // Default value if operand is out of range

```

Comment:
I'd prefer to add an error case here, rather than falling back to addition.  As a suggestion, you could add a `BIN_INVALID` value to the enum, and then print an error in the `BINARY_OP_A` case if this returned `BIN_INVALID`

Reply:
Resolved locally !